File: pygame/player.py
#!/usr/bin/env python3
import pygame

from settings import *

class Player(pygame.sprite.Sprite):
    # =========================================================
    def __init__(self):
        pygame.sprite.Sprite.__init__(self)
        self.bird_images =  [
            pygame.image.load(RESOURSE_BLUEBIRD_DOWNFLAP_PATH),
            pygame.image.load(RESOURSE_BLUEBIRD_MIDFLAP_PATH),
            pygame.image.load(RESOURSE_BLUEBIRD_UPFLAP_PATH)
        ]
        self.image = self.bird_images[0]
        self.rect = self.image.get_rect()
        self.rect.center = BIRD_START_POS
        self.image_animation_index = 0
        self.is_started = False
        self.velocity = 0.0
        self.MAX_CENTER_Y_VAL = WINDOW_HEIGHT - RESOURCE_BIRD_HEIGHT
    
    # =========================================================
    def update(self, window):
        # Animate bird
        self.image_animation_index += 1
        if self.image_animation_index >= 20:
            self.image_animation_index = 0
        animation_index = self.image_animation_index // 5
        if(animation_index in [0, 1, 2]):
            self.image = self.bird_images[animation_index]
        else:
            self.image = self.bird_images[1]

        if self.is_started == True:
            self.move(window)

    # =========================================================
    def move(self, window):
        self.velocity += BIRD_ACCELERATION 
        if self.velocity >= BIRD_MAX_VELOCITY:
            self.velocity = BIRD_MAX_VELOCITY
        self.rect.centery += self.velocity

        if self.rect.centery <= 0:
            self.rect.centery = 0

        if self.rect.centery >= self.MAX_CENTER_Y_VAL:
            self.rect.centery = self.MAX_CENTER_Y_VAL


    # =========================================================
    def on_flap(self):
        self.velocity = BIRD_FLAP_VELOCITY

    # =========================================================
    def on_start(self):
        self.is_started = True
        self.velocity = 0.0

    # =========================================================
    def on_game_over(self):
        self.is_started = False
        self.velocity = 0.0

File: sfml/src/Player.cpp
#include "Player.h"

#include "settings.h"

Player :: Player() : m_textures(), m_sprite(),
    m_animationCounter(0), m_started(false), m_velocity(0.0f)
{
    if (!m_textures[0].loadFromFile(RESOURSE_BLUEBIRD_DOWNFLAP_PATH) ||
        !m_textures[1].loadFromFile(RESOURSE_BLUEBIRD_MIDFLAP_PATH) ||
        !m_textures[2].loadFromFile(RESOURSE_BLUEBIRD_UPFLAP_PATH))
    {
        throw std::runtime_error("Failed to load bird textures");
    }

    m_sprite.setTexture(m_textures[0]);

    auto bounds = m_sprite.getLocalBounds();
    m_sprite.setOrigin(
        bounds.width / 2.f,
        bounds.height / 2.f);

    m_sprite.setPosition(BIRD_START_X, BIRD_START_Y);
}

void Player :: update(const sf::RenderWindow& window)
{
    animate();

    if (m_started)
    {
        move(window);
    }
}

void Player :: draw(sf::RenderWindow& window)
{
    window.draw(m_sprite);
}

void Player :: onFlap()
{
    m_velocity = BIRD_FLAP_VELOCITY;
}

void Player :: onStart()
{
    m_started = true;
    m_velocity = 0.0f;
}

void Player :: onGameOver()
{
    m_started = false;
    m_velocity = 0.0f;
}

void Player :: animate()
{
    ++m_animationCounter;

    if (m_animationCounter >= 20)
    {
        m_animationCounter = 0;
    }

    int animationIndex = m_animationCounter / 5;

    if (animationIndex <= 2)
    {
        m_sprite.setTexture(m_textures[animationIndex], true);
    }
    else
    {
        m_sprite.setTexture(m_textures[1], true);
    }
}

void Player :: move(const sf::RenderWindow& window)
{
    m_velocity += BIRD_ACCELERATION;

    if (m_velocity > BIRD_MAX_VELOCITY)
    {
        m_velocity = BIRD_MAX_VELOCITY;
    }

    m_sprite.move(0.f, m_velocity);

    float y = m_sprite.getPosition().y;

    float maxY =
        static_cast<float>(window.getSize().y)
        - RESOURCE_BIRD_HEIGHT;

    y = std::clamp(y, 0.f, maxY);

    m_sprite.setPosition(
        m_sprite.getPosition().x,
        y);

    m_sprite.setRotation(
        m_velocity * BIRD_ANGULAR_SPEED);
}